sql/src/main/java/io/crate/metadata/functions/params/FuncParams.java

```diff
@@ -0,0 +1,275 @@
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Type handling for function parameters (FuncParams).
+ *
+ * A FuncParam is composed of zero, one, or multiple parameters ({@link Param}s).
+ *
+ * Parameters are linked with each other by passing the same references more than once.
+ *
+ * For example:
+ * {@code
+ *
+ *  Param a = Param.of(ANY);
+ *  Param b = Param.of(INTEGER);
+ *
+ *  FuncParams.of(a, a, b).build();
+ *  // => A function parameters definition with three parameters.
+ *  // The first and the second parameter (a) can be of any type but must have the same type.
+ *  // The third parameter (b) has to be of type integer.
+ * }
+ *
+ * Matching
+ *
+ * Parameters are matched against an argument list by using the match method:
+ * {@code
+ *
+ *  List<Symbol> args = Arrays.asList(Literal.of(1L), Literal.of(1), Literal.of(1))
+ *
+ *  List<DataType> signature = FuncParams.match(args);
+ *  // => The resulting signature does not have to be of the same type as the original symbols
+ *  // but the implementation assures that a valid conversion is derived from the given types.
+ *  // This is done by checking convertibility and type precedence.
+ * }
+ *
+ * Variable Arguments (VarArgs)
+ *
+ * Additional variable arguments can be specified on FuncParams.Builder:
+ * {@code
+ *
+ *      FuncParams.of().withVarArgs(Param.of(INTEGER)).build()
+ *      // => Any number of integers as var args.
+ *      FuncParams.of().withVarArgs(Param.of(STRING), Param.of(INTEGER)).build()
+ *      // => Any number of String followed by an Integer parameter,
+ *      // e.g. "a", 1, "b", 2, ...
+ *
+ *      FuncParams.of().withVarArgs(2, Param.of(INTEGER)).build()
+ *      // => Maximum of two variable integer arguments.
+ *
+ *      FuncParams.of().withIndependentVarArgs(Param.of(ANY)).build()
+ *      // => Unlimited number of arguments which don't have to be of the same type
+ * }
+ *
+ *
+ * This class is thread-safe and can be accessed by multiple threads at once.
+ */
+public class FuncParams {
+
+    public static final FuncParams NONE = new FuncParams();
+    public static final FuncParams SINGLE_ANY = FuncParams.of(Param.ANY).build();
+    public static final FuncParams SINGLE_NUMERIC = FuncParams.of(Param.NUMERIC).build();
+    public static final FuncParams ANY_VAR_ARGS_SAME_TYPE = FuncParams.of().withVarArgs(Param.ANY).build();
+
+    private final Param[] parameters;
+
+    private Param[] varArgParameters;
+    private int maxVarArgOccurrences;
+    private boolean varArgsIndependent;
+
+    private FuncParams(Param... parameters) {
+        this.parameters = parameters;
+        this.varArgParameters = new Param[]{};
+        this.maxVarArgOccurrences = -1;
+    }
+
+    public static class Builder {
+
+        private FuncParams funcParams;
+
+        public Builder(Param... parameters) {
+            this.funcParams = new FuncParams(parameters);
```

Comment:
It looks like the builder is mostly used during node-startup and not at runtime - wouldn't it be cleaner to make all FuncParams attributes final and have the Builder have it's own mutable attributes?

Reply:
Or just provide the appropriate constructors in `FuncParams` - it look like the methods are never really chained?

Reply:
> The Builder doesn't cause runtime overhead. This is a really minor issue. I'd rather keep the API readable for other people then optimizing one object allocation upon startup.

That's kinda my point - I'd be in favor of *not* optimizing the object allocation and properly decouple the Builder attributes from the FuncParams and then create the FuncParams with all fields final.

If you're reading code and try to figure out if something really *is* thread-safe, it's much easier if the attributes are marked as final vs. "effectively" final.